File: DynamicsAndControl/GoodVibrationsWithFreeball/TheFiniteElementMethod/CPP/src/03_bar_element.cpp
#include "TheFiniteElementMethod/03_bar_element.h"

BarElement::BarElement()
{
}

BarElement::~BarElement()
{

}

void BarElement::RemoveRow(Eigen::MatrixXd& matrix, unsigned int rowToRemove)
{
  unsigned int numRows = matrix.rows()-1;
  unsigned int numCols = matrix.cols();

  if( rowToRemove < numRows )
    matrix.block(rowToRemove,0,numRows-rowToRemove,numCols) = matrix.bottomRows(numRows-rowToRemove);

  matrix.conservativeResize(numRows,numCols);
}

void BarElement::RemoveColumn(Eigen::MatrixXd& matrix, unsigned int colToRemove)
{
  unsigned int numRows = matrix.rows();
  unsigned int numCols = matrix.cols()-1;

  if( colToRemove < numCols )
    matrix.block(0,colToRemove,numRows,numCols-colToRemove) = matrix.rightCols(numCols-colToRemove);

  matrix.conservativeResize(numRows,numCols);
}

void BarElement::ReduceDimention(const int& dof, MatrixXd& M)
{
  M.block(dof, 0, M.rows() - 1, M.cols()) = M.bottomRows(M.rows() - 1);
  M.block(0, dof, M.rows(), M.cols() - 1) = M.rightCols(M.cols() - 1);
  M.conservativeResize(M.rows() - 1, M.cols() - 1);
}

std::tuple<MatrixXd, MatrixXd, VectorXd, MatrixXd> BarElement::Bar(const int num_elems)
{
  MatrixXd m, k;
  m.resize(2, 2);
  k.resize(2, 2);
  
  m << 2, 1,
        1, 2;

  k <<  1, -1,
        -1,  1;


  // element mass and stiffness matrices for a bar 
  m /= (6.0f * static_cast<float>(num_elems));
  k *= static_cast<float>(num_elems);

  // construct global mass and stiffness matrices
  MatrixXd M = MatrixXd::Zero(num_elems + 1, num_elems + 1);
  MatrixXd K = MatrixXd::Zero(num_elems + 1, num_elems + 1);

  // Assembly of elements
  for (int i = 0; i < num_elems; ++i)
  {
    M.block(i, i, 2, 2) += m;
    K.block(i, i, 2, 2) += k;
  }



  // Restrained degrees of freedom
  std::vector<int> restrained_dofs = {0};
  for(int dof : restrained_dofs)
  {
    ReduceDimention(dof, M);
    ReduceDimention(dof, K);
  }

  // Eigenvalue problem: Solve the generalized eigenvalue problem
  Eigen::GeneralizedSelfAdjointEigenSolver<MatrixXd> es(K, M);
  VectorXd evals = es.eigenvalues();
  MatrixXd evecs = es.eigenvectors();

  // Calculate frequencies
  VectorXd frequencies = evals.cwiseSqrt();

  return std::make_tuple(M, K, frequencies, evecs);
}

void BarElement::PlotFrequencies(const int n)
{
  double exact_frequency = M_PI / 2;
  // std::vector<std::pair<int, double>> results;
  std::vector<int> elements;
  std::vector<double> errors;

  for (int i = 1; i < n; ++i)
  {
    MatrixXd M, K;
    VectorXd frequencies;
    MatrixXd evecs;
    std::tie(M, K, frequencies, evecs) = Bar(i);

    double error = (frequencies(0) - exact_frequency) / exact_frequency * 100.0;
    // printf("Num Elems: %d, Fund. Frequency: %f, Error: %f\n", 
    //   i, frequencies[0], error);

    elements.emplace_back(i);
    errors.emplace_back(error);
  }

  plt::plot(elements, errors, "o-");
  plt::xlim(elements.at(0), elements.at(elements.size() - 1));
  plt::xlabel("Number of Elements");
  plt::ylabel("Error (%)");
  plt::show();
}

int main()
{
  BarElement be;
  be.PlotFrequencies(11);
  
  return 0;
}

File: DynamicsAndControl/GoodVibrationsWithFreeball/TheFiniteElementMethod/Python/03_bar_element.py
# https://www.youtube.com/watch?v=nGN9uqP67Ak&list=PL2ym2L69yzkaue8Ly2Oz51LALRzUV8LZ0&index=3

import numpy as np
import math

from scipy.linalg import eigh
from matplotlib import pyplot as plt

def Bar(num_elems):
  restrained_dofs = [0,]
  """  
  [m] = (p A l / 6) * 2   1
                      1   2


  [k] = E A /l * 1  -1
                -1   1
  """

  # element mass and stiffness matrices for a bar 
  # (2, 2)
  m = np.array([[2,  1], [1, 2]]) / (6. * float(num_elems))
  k = np.array([[1, -1], [-1, 1]]) * float(num_elems)

  # construct global mass and stiffness matrices
  # (num_elems + 1, num_elems + 1)
  M = np.zeros((num_elems + 1, num_elems + 1))
  K = np.zeros((num_elems + 1, num_elems + 1))

  # assembley of elements
  for i in range(num_elems):
    # print('num_elems: {} \ti: {}'.format(num_elems, i))
    #(num_elems + 1, num_elems + 1)
    M_temp = np.zeros((num_elems + 1, num_elems + 1))
    K_temp = np.zeros((num_elems + 1, num_elems + 1))

    M_temp[i : i + 2, i : i + 2] = m
    K_temp[i : i + 2, i : i + 2] = k

    M += M_temp
    K += K_temp

  # print("restrained_dofs")
  # print(restrained_dofs)
  # removed the fixed degree of freedom
  for dof in restrained_dofs:
    for i in [0, 1]:
      M = np.delete(M, dof, axis=i)
      K = np.delete(K, dof, axis=i)

  # eigen value problem
  evals, evecs = eigh(K, M)
  frequencies = np.sqrt(evals)

  return M, K, frequencies, evecs

def PlotFrequencies(n):
  exact_frequency = math.pi / 2
  results = []

  for i in range(1, n):
    M, K, frequencies, evecs = Bar(i)
    error = (frequencies[0] - exact_frequency) / exact_frequency * 100.0
    results.append((i, error))

    # print('Num Elems: {} \tFund. Frequency: {} \t Error: {}%'.format(i, round(frequencies[0], 3), round(error, 3)))


  # print('Exact frequency: ', round(exact_frequency, 3))

  # plot the results
  elements = np.array([x[0] for x in results])
  errors   = np.array([x[1] for x in results])

  plt.plot(elements, errors, 'o-')
  plt.xlim(elements[0], elements[-1])
  plt.xlabel('Number of Elements')
  plt.ylabel('Error (%)')
  plt.show()

if __name__ == '__main__':
  PlotFrequencies(11)
